xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -496,49 +577,143 @@ private <T extends ResourceUpdate> void handleResourceUpdate(
+      if (controlPlaneClient.equals(subscriber.controlPlaneClient)) {
         subscriber.onAbsent(processingTracker);
       }
     }
   }
 
+  /**
+   * Try to fallback to a lower priority control plane client.
+   *
+   * @param activeServerInfo  Matches the currently active control plane client
+   * @param serverInfos The full ordered list of xds servers
+   * @param authority The authority within which we are falling back
+   * @return true if a fallback was successful, false otherwise.
+   */
+  private boolean doFallbackForAuthority(
+      ServerInfo activeServerInfo, List<ServerInfo> serverInfos, String authority) {
+    if (serverInfos == null || serverInfos.size() < 2) {
+      return false;
+    }
+
+    // Build a list of servers with lower priority than the current server.
+    List<ServerInfo> fallbackServers = new ArrayList<>();
+    boolean found = false;
+    for (ServerInfo cur : serverInfos) {
+      if (cur.equals(activeServerInfo)) {
+        found = true;
+      } else if (found) {
+        fallbackServers.add(cur);
+      }
+    }
+
+    ControlPlaneClient fallbackCpc =
+        !fallbackServers.isEmpty()
+        ? getOrCreateControlPlaneClient(ImmutableList.copyOf(fallbackServers))
+        : null;
+
+    return fallBackToCpc(fallbackCpc, authority, activeServerInfo.target());
+  }
+
+  private boolean fallBackToCpc(
+      ControlPlaneClient fallbackCpc, String authority, String oldTarget) {
+    boolean didFallback = false;
+    if (fallbackCpc != null && ! fallbackCpc.isInBackoff()) {
+      logger.log(XdsLogLevel.INFO, "Falling back to XDS server {0}",
+          fallbackCpc.getServerInfo().target());
+
+      setCpcForAuthority(authority, fallbackCpc);
+      fallbackCpc.sendDiscoveryRequests(authority);
+      didFallback = true;
+    } else {
+      logger.log(XdsLogLevel.WARNING, "No working fallback XDS Servers found from {0}", oldTarget);
+    }
+    return didFallback;
+  }
+
+  private void doFallbackIfNecessary(ControlPlaneClient cpc, String authority) {
+    if (cpc == null) {
+      return;
+    }
+
+    ControlPlaneClient activeCpClient = activeCpClients.get(authority);
+    if (cpc == activeCpClient) {
+      return;
+    }
+    if (activeCpClient == null) {
+      setCpcForAuthority(authority, cpc);
+      return;
+    }
+
+    fallBackToCpc(cpc, authority, activeCpClient.getServerInfo().target());
+  }
+
+  private void setCpcForAuthority(String authority, ControlPlaneClient cpc) {
+    activeCpClients.put(authority, cpc);
+
+    for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> resourceSubscriberMap
+        : resourceSubscribers.values()) {
+      for (ResourceSubscriber<? extends ResourceUpdate> subscriber
+          : resourceSubscriberMap.values()) {
+        if (Objects.equals(subscriber.authority, authority)) {
+          ControlPlaneClient oldCpc = subscriber.controlPlaneClient;
+          subscriber.controlPlaneClient = cpc;
+          if (oldCpc != null && oldCpc != cpc) {
```

Comment:
we can keep the cached data in subscriber, no?
because once fall back, the data can be overridden. And during the fall back window, we can serve the cache data for the new watchers. In case the fallback is not successful and it flaps back to the original server, we do not introduce delays on the new watchers.

Reply:
You were right, updated.